### PR TITLE
Audio dataset pipeline

### PR DESCRIPTION
**Pull request by owengc**
- audio DSP for feature extraction and corresponding tests
- functions for working with Spoken Digit audio dataset

### CODE REVIEW

File: datasets.py [hunk 1]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+'''
+Spoken digit audio dataset
+'''
+
+def _preprocess_spoken_digit_audio_tf(example, with_labels):
+    audio = example['audio']
+    # convert from int16? to float32 [-1.0, 1.0]
+    if audio.dtype.is_integer:
+        # detected clipping using just tf.int16.max to normalize...
+        audio = tf.cast(audio, tf.float32) / (tf.int16.max + 1.0)
```
Comment (payamr) [inline datasets.py]: Normalization should not depend on sample. If we truly need to mix data coming from different sources, such that they'd have different types and ranges, we'd have a lot more to worry about such as differences in distributions. For now, let's assume that all data has the same type and range. More on what to do in the comment below.

File: datasets.py [hunk 2]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+        
+    # remove DC offset
+    audio_mean = tf.math.reduce_mean(audio, axis=0) 
+    audio -= audio_mean
```
Comment (payamr) [inline datasets.py:51]: Data normalization should not be done with the mean value of the given example. Any kind of normalization should be done with the same parameters for all examples. These parameters could be either fixed or could come from the statistics of your dataset. Common cases:

- fixed: for positive-value data, normalize with max and then center around zero, i.e. x_norm = (x / x_max) - 0.5
- standardize: for any data, x_norm = (x - x_mean) / x_std, where mean and std are taken over your entire dataset.

We can go with the first one here which is simpler to calculate. I'd add a parameter to the function with a default value to represent x_max, say `max_value=2**16`.

File: datasets.py [hunk 3]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
     test_dataset = tf.data.Dataset.from_tensor_slices(test_images).shuffle(test_size).batch(batch_size)
     return train_dataset, test_dataset
 
+'''
```
Comment (payamr) [inline datasets.py:41]: General comment: it'd be great to add `typing` and docstrings to all functions.

File: datasets.py [hunk 4]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+    all_data = tfds.load(name='spoken_digit',
+                         split=[f'train[:{percent_load}%]'],
+                         shuffle_files=False,
+                         as_supervised=False,
+                         download=download)
+
+    all_data = all_data[0]
+    all_data = all_data.map(lambda x: _preprocess_spoken_digit_audio_tf(x, load_labels), 
+                            num_parallel_calls=tf.data.AUTOTUNE,
+                            deterministic=False)
```
Comment (payamr) [inline datasets.py]: `deterministic` should not be set to False, because the order impacts what becomes train or test downstream. We want to be able to get a reproducible set which we can shuffle individually.

File: datasets.py [hunk 5]
```diff
@@ -1,9 +1,16 @@
 from typing import *
+from pathlib import Path
```
Comment (payamr) [inline datasets.py:2]: Can we just use `os.path` so we won't need an extra library?

File: datasets.py [hunk 6]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+                         download=download)
+
+    all_data = all_data[0]
+    all_data = all_data.map(lambda x: _preprocess_spoken_digit_audio_tf(x, load_labels), 
+                            num_parallel_calls=tf.data.AUTOTUNE,
+                            deterministic=False)
+    return all_data
+
+
+def convert_raw_spoken_digit_dataset(destination_dir, percent_load=100, download=False):
```
Comment (payamr) [inline datasets.py:77]: Why is `download` set to `False` by default?

File: datasets.py [hunk 7]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+    
+    if label is not None:
+        label = tf.convert_to_tensor(label, dtype=tf.int8)
+        return (features, label)
+    
+    return features
+
+
+def _count_spoken_digit_dataset_npy(data_dir):
+    return sum([1 for _ in data_dir.glob('*.npy')])
```
Comment (payamr) [inline datasets.py:155]: `len(data_dir.glob('*.npy')`?

File: datasets.py [hunk 8]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+        load_labels=False,
+        repeat=False,
+) -> Tuple[tf.data.Dataset, tf.data.Dataset]:    
+    all_data, padded_shapes = load_spoken_digit_dataset(data_dir, config, load_labels)
+    total_count = len(all_data)
+
+    train_count = int(train_test_split * total_count)
+    test_count = total_count - train_count
+    
+    train_data = all_data.take(train_count).cache().shuffle(train_count, reshuffle_each_iteration=True)\
```
Comment (payamr) [inline datasets.py]: The `.take/ .skip` way works, but there are flaws with it. Notably, when the data is sorted in a particular way, say by speaker or by digit in our case, we'll end up with a test set which may consist only a subset of digits or speakers, while that same digit or speaker would be completely missing from the train set. Note that this is also because we don't want to shuffle before splitting, especially when the shuffle isn't reproducible. 

I think a good way here would be to allow `load_spoken_digit_dataset` and `_load_spoken_digit_dataset_npy` to generate test and train datasets. For example, it could iterate over files in `data_dir` and take every other `n` samples for the test set, and the rest for train. This is what I had done with the `skip_condition` in the code you have removed.

File: datasets.py [hunk 9]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+    
+    if repeat == False:
```
Comment (payamr) [inline datasets.py]: There are no realistic cases where we'd not want to repeat the dataset when we're creating from a generator. Also the assert below, if ever run, would take a very long time because it essentially would have to iterate over the entire object. I'd remove this, but if you want to keep it I'd set the default `repeat=True`. Also below it seems you've used `repeat` as an int which makes the typing too flexible. Just don't set the `count` in `repeat`, the model trainer will handle things.

File: datasets.py (hunk 8 above)
Comment (payamr) [reply datasets.py]: Also one motivation for saving things as numpy files was to provide more flexibility.

File: datasets.py [hunk 10]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+    signature = [signature]
+        
+    if load_labels:
+        signature.append(tf.TensorSpec(shape=(), dtype=tf.int8))
+        padded_shapes = (padded_shapes, ())
+    
+    return tuple(signature), padded_shapes
+
+
+def _extract_spoken_digit_features_npy(example, config, load_labels):
```
Comment (payamr) [inline datasets.py:120]: For this and other functions below, maybe except for the top level one: generally not a fan of passing a generic `config` while the function makes assumptions about what it expects to exist in the config. Each function seems to use 2 or 3 specific values, so I'd just pass those as explicit parameters.

File: datasets.py [hunk 11]
```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
...
+    test_count = total_count - train_count
+    
+    train_data = all_data.take(train_count).cache().shuffle(train_count, reshuffle_each_iteration=True)\
+    
+    if repeat == False:
+        assert(len(train_data) == train_count)
+    else:
+        train_data = train_data.repeat(count=repeat)
+    test_data = all_data.skip(train_count).cache()
+    assert(len(test_data) == test_count)
```
Comment (payamr) [inline datasets.py]: Remove this assert too. Also you want to add a repeat to the test data as well.

File: dsp/dsp.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+def compute_cola_window(config):
+    """
+    Using 'window_type' and 'window_size' values from 'config', computes 
+    desired window and corresponding hop size to achieve constant-overlap-add 
+    (COLA) property. Note that with some window - window size combinations the
+    quality of the COLA reconstruction is not perfect.
+    """
+    window_type = config['window_type']
+    window_size = config['window_size']
+    assert(window_size >= 64 and (window_size & (window_size - 1)) == 0)
```
Comment (payamr) [inline dsp/dsp.py:20]: Not sure why `(window_size & (window_size - 1)) == 0` is needed.

File: dsp/dsp.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+window_types = [
+    'boxcar',
+    'bartlett',
+    'hann',
+    'blackman',
+    'blackmanharris'
+]
+window_sizes = [64, 128, 256, 512, 1024, 2048, 4096]
+
+def compute_cola_window(config):
```
Comment (payamr) [inline dsp/dsp.py]: Similar to the other file, can we explicitly pass `window_type` and `window_size`?

File: dsp/dsp.py [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+    return window, hop_size
+
+def get_windowed_frame_size(config):
+    window_size = config['window_size']
+    padding_factor = config['padding_factor']
+    frame_size = window_size * pow(2, padding_factor)
+    return frame_size
+
+def windowed_frame_analysis(signal, config):
+    window = config['window']
```
Comment (payamr) [inline dsp/dsp.py]: I didn't see a `window` in the sample config file, but maybe I'm missing it. Which brings me to my main point which is that it seems that a lot of these asserts can be collected in a `validate` function which would validate things once after the config is loaded. Some of this code is repeated in the next functions.

File: dsp/dsp.py [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+    return signal
+
+def get_fft_bin_phase_frequencies(num_bins, hop_size, sample_rate):
+    bin_frequencies = np.linspace(0, sample_rate // 2, num=num_bins, endpoint=True)
+    bin_phases = np.linspace(0, np.pi * hop_size, num=num_bins, endpoint=True)
+    return bin_frequencies, bin_phases
+
+def get_feature_diffs(feature_frames):
+    feature_diff_frames = np.vstack((feature_frames[0], np.diff(feature_frames, axis=0)))
+    assert(np.allclose(feature_frames.shape, feature_diff_frames.shape))
```
Comment (payamr) [inline dsp/dsp.py]: Is this assert really needed?

File: dsp/dsp.py [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+    return feature_diff_frames
+
+def get_feature_cumsum(feature_diff_frames):
+    feature_frames = np.cumsum(feature_diff_frames, axis=0)
+    assert(np.allclose(feature_frames.shape, feature_diff_frames.shape))
```
Comment (payamr) [inline dsp/dsp.py]: Is this assert really needed?

File: dsp/dsp.py [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+        n_end = n_start + window_size
+        signal[n_start : n_end] += windowed_frames[i]
+    return signal
+
+def get_fft_bin_phase_frequencies(num_bins, hop_size, sample_rate):
+    bin_frequencies = np.linspace(0, sample_rate // 2, num=num_bins, endpoint=True)
+    bin_phases = np.linspace(0, np.pi * hop_size, num=num_bins, endpoint=True)
+    return bin_frequencies, bin_phases
+
+def get_feature_diffs(feature_frames):
```
Comment (payamr) [inline dsp/dsp.py:160]: An implicit assumption in this and other functions here is that `time` corresponds to `axis=0`. Maybe document this.

File: dsp/dsp.py [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+    num_hops = (N - window_size) // H + 1
+    M = num_hops * H + window_size
+    if M > N:
+        signal = np.append(signal, np.zeros(M - N))
+    assert(len(signal) == M)
+    
+    num_hops = (M - window_size) // H + 1
+    
+    if sqrt_window and np.all(window > 0):
+        window = np.sqrt(window)
```
Comment (payamr) [inline dsp/dsp.py:96]: Shouldn't this raise an exception when sqrt_window is true but some values are negative?

File: datasets.py (hunk 1 above)
Comment (owengc) [reply datasets.py]: The things I'm doing here are audio-specific and shouldn't be construed as the kind of data standardization you're referring to. While it's true that we may eventually want to use this code with data that's encoded in something other than 16-bit I was under the impression that, like the other functions for the image dataset, we're tailoring the pipeline to this particular set of files for now, which IS stored as 16-bit. 

This particular block is for converting audio data stored as integers to floating point [-1, 1], which is how we want to actually work with it. More below

File: datasets.py (hunk 2 above)
Comment (owengc) [reply datasets.py:51]: Again, though this kind of looks like the sort of standardization you're referring to, it isn't. What I'm doing here is removing the DC offset of the sample. When we have the entire sample available this is the easiest method of doing so and *must* be done using the mean of each individual sample. Using the mean of the entire dataset wouldn't accomplish the same goal, it would just shift the DC offset to some other value. 

If we were building a realtime system which works on short blocks of audio (rather than capturing a complete utterance / phrase), I would instead use a DC-blocking filter. But since we're performing inference on these clips which do contain the entire signal of interest, we can safely use this simpler approach. 

The reason I'm doing this stuff here is that when we load the .npy data we don't want to have to do it over again. It's not discarding any useful information or making any unreasonable assumptions about the nature of the input data. We just need floating point audio with minimal DC offset in the range (-1, 1), and that's not what we get from the raw TF dataset so I'm converting it here before we write to the .npy files so we don't have to worry about doing it later. 

The stuff you're talking about might apply to some of the features we later compute from the raw audio, but that's at a different point in the pipeline and so far the features we're considering are already constrained to reasonable ranges.

File: datasets.py (hunk 3 above)
Comment (owengc) [reply datasets.py:41]: yeah, agreed. I can do that once the dust settles

Comment (owengc) [reply datasets.py:41]: (though I'm not actually sure how to indicate types for optional / conditional return values)

File: datasets.py (hunk 4 above)
Comment (owengc) [reply datasets.py]: OK, I'll change this

File: datasets.py (hunk 5 above)
Comment (owengc) [reply datasets.py:2]: Don't sleep on `Path`s. I much prefer using `pathlib` and `shutil` to `os`, they're both built-in libs just like `os` anyway

File: datasets.py (hunk 6 above)
Comment (owengc) [reply datasets.py:77]: My reasoning is that the user should be forced to do that consciously since it starts a long process / downloads a large file to their system

That said, once it's downloaded I don't think it re-downloads if it finds it already residing in the expected location, so it's not a huge deal either way.

File: datasets.py (hunk 7 above)
Comment (owengc) [reply datasets.py:155]: I think it would have to be `len(list(data_dir.glob('*.npy')))` unless `len(generator)` is some newer feature I don't know about..

I agree the syntax isn't beautiful but that's why I put it in a function. `len(list)` approach might be slightly less desirable from a memory standpoint in situations where there are many elements, but if you feel strongly about it I can change it, probably doesn't matter in this case

File: datasets.py (hunk 8 above)
Comment (owengc) [reply datasets.py]: I assume there's some mechanism in TF that would be better than putting the `if n % N` kind of thing in the generator. What if we just use a fixed random seed and shuffle the full dataset before the `.take` / `.skip` ?

File: datasets.py (hunk 9 above)
Comment (owengc) [reply datasets.py]: it's useful for debugging purposes to have a mode where the generator has a finite size, so I do want to keep the option there. 

About the `assert`s, they were helpful while I was writing the code and might help avoid some regressions but if you think it will really slow things down I don't mind removing them now that things are working as expected

File: datasets.py (hunk 11 above)
Comment (owengc) [reply datasets.py]: ok

File: datasets.py (hunk 10 above)
Comment (owengc) [reply datasets.py:120]: The reason I went for the generic config approach is that depending on the kinds of features we end up using we would have a lot of individual parameters and the API could change a lot. Doing it this way makes it easier to accommodate changes later without changing the API. 

In general I agree with the point you're trying to make, and in most situations I would agree with you, but this does streamline things a lot. Maybe we ought to have some sort of 'config validation' or a README or something if you're worried about it being too opaque. 

My though was that since there's this large collection of parameters, many of which are shared between the various functions, it's convenient to gradually build up all the necessary information and collect it in this object. We might start with a couple explicit arguments like 'window_type' and 'window_size' but from there a lot of stuff is computed automatically during the various steps from 'raw audio' to 'stft feature frames' and though I agree that it's confusing right now because it's not documented, letting those functions responsible for each step handle inserting the required information into the config object avoids a lot of clutter and forwarding along a bunch of values that kind of all need to be bundled together anyway.

At the moment it might be manageable to pass everything individually but again, this approach makes it easier to add / remove functionality as we try things without having to continuously change the API as we go.

File: dsp/dsp.py (hunk 13 above)
Comment (owengc) [reply dsp/dsp.py]: sure, I think that makes sense here. Originally I was adding the return values of this function to the config object, so it made sense to just pass in the params that way. So we could also put it back that way and just lean into the 'incrementally build up the config object' approach you're so fond of ;)

File: dsp/dsp.py (hunk 12 above)
Comment (owengc) [reply dsp/dsp.py:20]: we want `window_size` to be a power of two. this is a little cryptic but that's all it's doing. i can put a comment about it in the docstring if you'd like

File: dsp/dsp.py (hunk 14 above)
Comment (owengc) [reply dsp/dsp.py]: Right now it's the user's responsibility to add `window` and `hop_size` to the config. This kind of goes against the pattern I was using overall, so as I said above it might make sense to change it so that `compute_cola_window` automatically adds it to the config. However I wanted to leave the option open of using arbitrary, 'non-COLA' `window` / `hop_size` settings. There isn't a `window` entry in the config file because the `compute_cola_window` method generates it

you can think of these various DSP operations as nodes in a directed graph. We need more / different information as we traverse it to specify the behavior of each node, and by letting these functions consume/pass along that information via the config object we're simplifying the process of composing that graph. 

again, I can see arguments against this approach and I do agree that it needs to be documented, but it does streamline things

File: dsp/dsp.py (hunk 18 above)
Comment (owengc) [reply dsp/dsp.py:96]: yes it should probably raise an exception instead of failing silently, I'll add that

File: dsp/dsp.py (hunk 16 above)
Comment (owengc) [reply dsp/dsp.py]: no, I meant to move things like this to the tests. i'll do that and remove them here

File: datasets.py (hunk 8 above)
Comment (owengc) [reply datasets.py]: i've added a deterministic shuffle to the raw TF dataset prior to writing the examples as numpy files
